app/models/spree/payment_decorator.rb

```diff
@@ -1,9 +1,11 @@
 Spree::Payment.class_eval do
-  before_update :send_gift_card, :if => "state_changed? && state_was != 'completed' && state == 'completed'"
+  after_update :send_gift_card
 
   def send_gift_card
-    order.line_items.each do |li|
-      Spree::OrderMailer.gift_card_email(li.gift_card.id, order).deliver if li.gift_card
+    if state_changed? && state_was != 'completed' && state == 'completed'
```

Comment:
@AnuragBhardwaj1 Please move this condition to a separate method and use it with after_update